File: src/stages/stage1.ts
import path from "node:path";

import { runWithValidation } from "../agents/index.js";
import { CheckpointManager } from "../checkpoint.js";
import type { AuditConfig } from "../config.js";
import { getLogger } from "../logger.js";
import { loadPrompt } from "../prompts.js";
import { validateStage1File } from "../validation/stage1.js";

const logger = getLogger("stage1");
const TASK_KEY = "stage1";

export async function runStage1(
  config: AuditConfig,
  checkpoint: CheckpointManager,
): Promise<string> {
  const outputPath = path.join(config.outputDir, "stage-1-research.md");

  if (checkpoint.isComplete(TASK_KEY)) {
    logger.info("Stage 1 already complete, loading existing output.");
    return outputPath;
  }

  const today = new Date().toISOString().slice(0, 10);
  const startDate = new Date(Date.now() - 5 * 365.25 * 24 * 60 * 60 * 1000)
    .toISOString()
    .slice(0, 10);

  const prompt = await loadPrompt("stage1.md", {
    target_path: config.target,
    output_path: outputPath,
    today,
    start_date: startDate,
    user_instructions: config.scope || "No additional scope constraints.",
  });

  const { passed } = await runWithValidation({
    config,
    prompt,
    cwd: config.target,
    outputPath,
    validator: validateStage1File,
  });

  if (!passed) {
    logger.warning("Stage 1 validation did not fully pass, continuing with best-effort output.");
  }

  checkpoint.markComplete(TASK_KEY);
  logger.info("Stage 1 complete. Research report: %s", outputPath);
  return outputPath;
}

File: protocol_auditor/stages/stage1.py
from __future__ import annotations

import os

from ..agent import run_with_validation
from ..checkpoint import CheckpointManager
from ..config import AuditConfig, Module
from ..logger import get_logger
from ..parsing.stage1 import parse_modules
from ..prompts import load_prompt
from ..validation.stage1 import validate_stage1_file

logger = get_logger("stage1")
_TASK_KEY = "stage1"


async def run_stage1(config: AuditConfig, checkpoint: CheckpointManager) -> list[Module]:
    output_path = os.path.join(config.output_dir, "stage-1-modules.json")

    if checkpoint.is_complete(_TASK_KEY):
        logger.info("Stage 1 already complete, loading existing output.")
        return parse_modules(output_path)

    prompt = load_prompt("stage1.md", {
        "target_path": config.target,
        "output_path": output_path,
        "user_instructions": config.scope or "No additional scope constraints.",
    })

    passed, _ = await run_with_validation(
        prompt=prompt,
        config=config,
        cwd=config.target,
        output_path=output_path,
        validator=validate_stage1_file,
    )

    if not passed:
        logger.warning("Stage 1 validation did not fully pass, continuing with best-effort output.")

    checkpoint.mark_complete(_TASK_KEY)
    modules = parse_modules(output_path)
    logger.info("Stage 1 complete. Modules: %s", ", ".join(m.id for m in modules))
    return modules
